cxl/lib/libcxl.c

```diff
@@ -8756,3 +8756,612 @@ CXL_EXPORT int cxl_memdev_fbist_test_randomsequence(struct cxl_memdev *memdev,
+	// OPL size
+	for(int i=0; i<32; i++){
+		fprintf(stdout, "Trigger Config [%d]: %x\n", i, le32_to_cpu(hct_get_config_out->trig_config[i]));
+	}
+
+out:
+	cxl_cmd_unref(cmd);
+	return rc;
+	return 0;
+}
+
+
+#define CXL_MEM_COMMAND_ID_HCT_READ_BUFFER CXL_MEM_COMMAND_ID_RAW
+#define CXL_MEM_COMMAND_ID_HCT_READ_BUFFER_OPCODE 50693
+#define CXL_MEM_COMMAND_ID_HCT_READ_BUFFER_PAYLOAD_IN_SIZE 2
+#define CXL_MEM_COMMAND_ID_HCT_READ_BUFFER_PAYLOAD_OUT_SIZE 1024
+
+struct cxl_mbox_hct_read_buffer_in {
+	u8 hct_inst;
+	u8 num_entries_to_read;
+}  __attribute__((packed));
+
+struct cxl_mbox_hct_read_buffer_out {
+	u8 buf_end;
+	u8 num_buf_entries;
+	__le16 rsvd;
+	u8 buf_entry[1024];
+}  __attribute__((packed));
+
+CXL_EXPORT int cxl_memdev_hct_read_buffer(struct cxl_memdev *memdev,
+	u8 hct_inst, u8 num_entries_to_read)
+{
+	u8 *buf_out;
+	struct cxl_cmd *cmd;
+	struct cxl_mem_query_commands *query;
+	struct cxl_command_info *cinfo;
+	struct cxl_mbox_hct_read_buffer_in *hct_read_buffer_in;
+	struct cxl_mbox_hct_read_buffer_out *hct_read_buffer_out;
+	int rc = 0;
+
+	cmd = cxl_cmd_new_raw(memdev, CXL_MEM_COMMAND_ID_HCT_READ_BUFFER_OPCODE);
+	if (!cmd) {
+		fprintf(stderr, "%s: cxl_cmd_new_raw returned Null output\n",
+				cxl_memdev_get_devname(memdev));
+		return -ENOMEM;
+	}
+
+	query = cmd->query_cmd;
+	cinfo = &query->commands[cmd->query_idx];
+
+	/* update payload size */
+	cinfo->size_in = CXL_MEM_COMMAND_ID_HCT_READ_BUFFER_PAYLOAD_IN_SIZE;
+	if (cinfo->size_in > 0) {
+		 cmd->input_payload = calloc(1, cinfo->size_in);
+		if (!cmd->input_payload)
+			return -ENOMEM;
+		cmd->send_cmd->in.payload = (u64)cmd->input_payload;
+		cmd->send_cmd->in.size = cinfo->size_in;
+	}
+
+	hct_read_buffer_in = (void *) cmd->send_cmd->in.payload;
+
+	hct_read_buffer_in->hct_inst = hct_inst;
+	hct_read_buffer_in->num_entries_to_read = num_entries_to_read;
+	rc = cxl_cmd_submit(cmd);
+	if (rc < 0) {
+		fprintf(stderr, "%s: cmd submission failed: %d (%s)\n",
+				cxl_memdev_get_devname(memdev), rc, strerror(-rc));
+		 goto out;
+	}
+
+	rc = cxl_cmd_get_mbox_status(cmd);
+	if (rc != 0) {
+		fprintf(stderr, "%s: firmware status: %d\n",
+				cxl_memdev_get_devname(memdev), rc);
+		rc = -ENXIO;
+		goto out;
+	}
+
+	if (cmd->send_cmd->id != CXL_MEM_COMMAND_ID_HCT_READ_BUFFER) {
+		 fprintf(stderr, "%s: invalid command id 0x%x (expecting 0x%x)\n",
+				cxl_memdev_get_devname(memdev), cmd->send_cmd->id, CXL_MEM_COMMAND_ID_HCT_READ_BUFFER);
+		return -EINVAL;
+	}
+
+	hct_read_buffer_out = (void *)cmd->send_cmd->out.payload;
+	fprintf(stdout, "========================== read hif/cxl trace buffer ===========================\n");
+	fprintf(stdout, "Buffer End Reached: %x\n", hct_read_buffer_out->buf_end);
+	fprintf(stdout, "Number of buffer entries: %x\n", hct_read_buffer_out->num_buf_entries);
+
+	buf_out = (u8*) cmd->send_cmd->out.payload;
+	for(int i=4; i<cmd->send_cmd->out.size; i++){
+		fprintf(stdout, "Buffer Entry: %x\n", buf_out[i]);
```

Comment:
The output payload uses 4 byte values, but here you are printing 1 byte at a time. Also writing these to multiple lines prepended with "Buffer Entry:" is a bit confusing since all of them are part of a single buffer entry.

Reply:
Unresolved this comment as I think it was misunderstood, and changes made will still print 1 byte at a time. Let's do one of two options:
1. If you want to skip parsing for now and just print a raw dump of all of the buffer entries, change 9008 to fprintf(stdout, "%02x ", buf_out[i]); and then print a single newline after the for loop. This may be fine for now if this is only being used for validation automation and we're not concerned about human readability.
2. If we do want to parse the buffer entries and ensure they're sensibly printed, we can hop on a call and I can explain how that would be done, but here's a quick psuedocode:
buffer_entry_length = (send_cmd->out.size - 4) / num_buf_entries;
for (int i = 0; i < num_buf_entries; i++) {
  fprintf(stdout, "Buffer entry %d:\n", i);
  if (buffer_entry_length == 32) {
    // parse according to HIF buffer entry data
  }
  else if (buffer_entry_length == 72) {
    // parse according to FLIT buffer entry data
  }
  else {
  // We shouldn't hit this block, print a debug message and figure out what went wrong
}